test/test_util.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/gruntwork-io/terratest/modules/aws"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/retry"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	test_structure "github.com/gruntwork-io/terratest/modules/test-structure"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// define struct for each object Test Case
+type ObjectTestCase struct {
+	key             string
+	encryption      string
+	expectPassRead  bool
+	expectPassWrite bool
+}
+
+// define struct for each bucket Test Case
+type BucketTestCase struct {
+	testName     string
+	vars         map[string]interface{}
+	objTestCases []ObjectTestCase
+}
+
+// CondCreateObject conditionally creates an S3 Object inside `bucket` with `body` content
+// It is called when we want to test ReadOnly path
+// Creates object with default AWS session that presumably has PutObject permission
+func CondCreateObject(t *testing.T, awsRegion string, bucket string, body string, obj ObjectTestCase) {
+	// When encryption isn't set, trying to upload with default AWS session would fail as well
+	// hence the check here
+	if obj.expectPassWrite || obj.encryption == "" {
+		return
+	}
+
+	sess, err := aws.NewAuthenticatedSession(awsRegion)
+	if err != nil {
+		assert.FailNow(t, "Failed in creating session")
+	}
+	uploader := s3manager.NewUploader(sess)
+
+	logger.Log(t, fmt.Sprintf("Default Credential: uploading object %s...", obj.key))
+
+	_, err = UploadObjectWithUploaderE(
+		awsRegion,
+		bucket,
+		obj.key,
+		obj.encryption,
+		body,
+		uploader,
+	)
+
+	require.NoError(t, err, "Error raised when trying to upload objects that would be tested in ReadOnly paths")
+
+	logger.Log(t, fmt.Sprintf("Uploaded object %s with default credential", obj.key))
+
+}
+
+func UploadObjectWithUploaderE(awsRegion string, bucketName string, key string, encryption string, body string, uploader *s3manager.Uploader) (*s3manager.UploadOutput, error) {
+	s3Input := &s3manager.UploadInput{
+		Bucket: &bucketName,
+		Key:    &key,
+		Body:   strings.NewReader(body),
+	}
+
+	if encryption != "" {
+		s3Input.ServerSideEncryption = &encryption
+	}
+
+	return uploader.Upload(s3Input)
+
+}
+
+func GetS3ObjectWithSessionE(t *testing.T, bucket string, key string, sess *session.Session) (string, error) {
+	s3Client := s3.New(sess)
+
+	res, err := s3Client.GetObject(&s3.GetObjectInput{
+		Bucket: &bucket,
+		Key:    &key,
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(res.Body)
+	if err != nil {
+		return "", err
+	}
+
+	contents := buf.String()
+	logger.Log(t, fmt.Sprintf("Read contents from s3://%s/%s", bucket, key))
+
+	return contents, nil
+}
+
+func validatePutObject(t *testing.T, awsRegion string, bucket string, obj ObjectTestCase, body string, sess *session.Session) {
+	_, err := retry.DoWithRetryInterfaceE(t,
```

Comment:
What motivated you to add retries here? Was there an intermittent issue that you ran into? Why 4 tries with the 3 second sleep?

Reply:
This makes sure IAM policies were applied to the role, since IAM updates are not automatically reflected. (Terraform will create the role, create the policy and attach the policy only a few secs away from trying to PutObject).
Few links on this.
https://stackoverflow.com/questions/20156043/how-long-should-i-wait-after-applying-an-aws-iam-policy-before-it-is-valid
https://forums.aws.amazon.com/post!reply.jspa?messageID=196178

Reply:
Got it! Did you know about this ahead of time or something you ran into during testing?

Regardless, let's include it in a comment so that we capture why it's here.

Reply:
I've had problems in the past with this IAM inconsistency yes. Regarding this code, I did not have that in mind when first writing it, but caught a situation where that happened (a permissions error when it shouldn't have happened). So I added the retries. Both here and when assuming role.